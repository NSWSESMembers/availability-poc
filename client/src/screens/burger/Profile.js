import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { ActivityIndicator, Text, View, Image } from 'react-native';
import { connect } from 'react-redux';
import { graphql, compose } from 'react-apollo';
import md5 from 'md5';
import Prompt from 'react-native-prompt';

import Icon from 'react-native-vector-icons/FontAwesome';
import { extendAppStyleSheet } from '../style-sheet';
import CURRENT_USER_QUERY from '../../graphql/current-user.query';

import UPDATE_USERPROFILE_MUTATION from '../../graphql/update-userprofile.mutation';

const updateUserProfileMutation = graphql(UPDATE_USERPROFILE_MUTATION, {
  props: ({ mutate }) => ({
    updateUserProfile: ({ displayName }) =>
      mutate({
        variables: { user: { displayName } },
      }),
  }),
});

const styles = extendAppStyleSheet({
  container: {
    flex: 1,
  },
  username: {
    borderColor: '#777',
    borderBottomWidth: 1,
    borderTopWidth: 1,
    paddingHorizontal: 8,
    paddingBottom: 2,
    paddingTop: 5,
    fontSize: 14,
  },
  usernameHeader: {
    backgroundColor: '#dbdbdb',
    color: '#000000',
    paddingHorizontal: 8,
    paddingBottom: 2,
    paddingTop: 5,
    fontSize: 16,
  },
  email: {
    borderColor: '#777',
    borderBottomWidth: 1,
    borderTopWidth: 1,
    paddingHorizontal: 8,
    paddingBottom: 2,
    paddingTop: 5,
    fontSize: 14,
  },
  emailHeader: {
    backgroundColor: '#dbdbdb',
    color: '#000000',
    paddingHorizontal: 8,
    paddingBottom: 2,
    paddingTop: 5,
    fontSize: 16,
  },
  imageContainer: {
    paddingRight: 20,
    alignItems: 'center',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
  inputBorder: {
    flexDirection: 'row',
    borderColor: '#dbdbdb',
    borderBottomWidth: 1,
    borderTopWidth: 1,
    paddingVertical: 8,
    backgroundColor: '#ed3434',
  },
  inputInstructions: {
    color: '#777',
    fontSize: 26,
    flex: 1,
  },
  userContainer: {
    paddingLeft: 16,
    backgroundColor: '#c6c0c0',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
  gravatar: {
    paddingHorizontal: 10,
  },
  userInner: {
    flexDirection: 'row',
    backgroundColor: '#dbdbdb',
    alignItems: 'center',
    paddingVertical: 8,
    paddingRight: 0,
  },
});

class Profile extends Component {
  static navigationOptions = {
    title: 'More',
    tabBarIcon: ({ tintColor }) => <Icon size={28} name="bars" color={tintColor} />,
  };

  state = {
    promptVisible: false,
  }

  updateDisplayName = (value) => {
    this.setState({ promptVisible: false });
    this.props
      .updateUserProfile({ displayName: value });
  }

  openPrompt = () => {
    this.setState({ promptVisible: true });
  };

    closePrompt = () => {
      this.setState({ promptVisible: false });
    };

    cancelPrompt = () => {
      this.setState({ promptVisible: false });
    };

    render() {
      const { loading, user } = this.props;

      // render loading placeholder while we fetch data
      if (loading || !user) {
        return (
          <View style={[styles.loading, styles.container]}>
            <ActivityIndicator />
          </View>
        );
      }

      return (
        <View style={styles.container}>
          <Prompt
            title="Change Display Name"
            placeholder={user.displayName}
            defaultValue={user.displayName}
            visible={this.state.promptVisible}
            onCancel={this.cancelPrompt}
            onSubmit={this.updateDisplayName}
          />
          <View style={styles.userContainer}>
            <View style={styles.userInner}>
              <View style={styles.gravatar}>
                <Image
                  style={{ borderRadius: 5, width: 50, height: 50, paddingHorizontal: 10 }}
                  source={{ uri: `https://www.gravatar.com/avatar/${md5(user.email)}?d=mm` }}
                />
              </View>
              <Text
                onPress={this.openPrompt}
                style={styles.inputInstructions}
              >{user.displayName}
              </Text>
            </View>
          </View>
          <Text style={styles.usernameHeader}>User Name</Text>
          <Text style={styles.username}>{user.username}</Text>
          <Text style={styles.emailHeader}>Email Address</Text>
          <Text style={styles.email}>{user.email}</Text>
        </View>
      );
    }
}

Profile.propTypes = {
  loading: PropTypes.bool,
  updateUserProfile: PropTypes.func,
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    displayName: PropTypes.string.isRequired,
  }),
};

const userQuery = graphql(CURRENT_USER_QUERY, {
  skip: ownProps => !ownProps.auth || !ownProps.auth.token,
  options: ({ auth }) => ({ variables: { id: auth.id }, fetchPolicy: 'cache-only' }),
  props: ({ data: { loading, user } }) => ({
    loading,
    user,
  }),
});

const mapStateToProps = ({ auth }) => ({
  auth,
});

export default compose(
  connect(mapStateToProps),
  userQuery,
  updateUserProfileMutation,
)(Profile);
