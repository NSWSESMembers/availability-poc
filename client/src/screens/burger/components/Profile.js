import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { ActivityIndicator, Text, View, Image } from 'react-native';
import Prompt from 'react-native-prompt';

import gravatar from '../../../selectors/gravatar';
import { Container } from '../../../components/Container';

const styles = {
  username: {
    borderColor: '#777',
    borderBottomWidth: 1,
    borderTopWidth: 1,
    paddingHorizontal: 8,
    paddingBottom: 2,
    paddingTop: 5,
    fontSize: 14,
  },
  usernameHeader: {
    backgroundColor: '#dbdbdb',
    color: '#000000',
    paddingHorizontal: 8,
    paddingBottom: 2,
    paddingTop: 5,
    fontSize: 16,
  },
  email: {
    borderColor: '#777',
    borderBottomWidth: 1,
    borderTopWidth: 1,
    paddingHorizontal: 8,
    paddingBottom: 2,
    paddingTop: 5,
    fontSize: 14,
  },
  emailHeader: {
    backgroundColor: '#dbdbdb',
    color: '#000000',
    paddingHorizontal: 8,
    paddingBottom: 2,
    paddingTop: 5,
    fontSize: 16,
  },
  imageContainer: {
    paddingRight: 20,
    alignItems: 'center',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
  inputBorder: {
    flexDirection: 'row',
    borderColor: '#dbdbdb',
    borderBottomWidth: 1,
    borderTopWidth: 1,
    paddingVertical: 8,
    backgroundColor: '#ed3434',
  },
  inputInstructions: {
    color: '#777',
    fontSize: 26,
    flex: 1,
  },
  userContainer: {
    paddingLeft: 16,
    backgroundColor: '#c6c0c0',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
  gravatar: {
    paddingHorizontal: 10,
  },
  userInner: {
    flexDirection: 'row',
    backgroundColor: '#dbdbdb',
    alignItems: 'center',
    paddingVertical: 8,
    paddingRight: 0,
  },
};

class Profile extends Component {
  state = {
    promptVisible: false,
  }

  updateDisplayName = (value) => {
    this.setState({ promptVisible: false });
    this.props.onChangeName(value);
  }

  openPrompt = () => {
    this.setState({ promptVisible: true });
  };

  closePrompt = () => {
    this.setState({ promptVisible: false });
  };

  render() {
    const { loading, user } = this.props;

    // render loading placeholder while we fetch data
    if (loading || !user) {
      return (
        <View style={[styles.loading, styles.container]}>
          <ActivityIndicator />
        </View>
      );
    }

    return (
      <Container>
        <Prompt
          title="Change Display Name"
          placeholder={user.displayName}
          defaultValue={user.displayName}
          visible={this.state.promptVisible}
          onCancel={this.closePrompt}
          onSubmit={this.updateDisplayName}
        />
        <View style={styles.userContainer}>
          <View style={styles.userInner}>
            <View style={styles.gravatar}>
              <Image
                style={{ borderRadius: 5, width: 50, height: 50, paddingHorizontal: 10 }}
                source={{ uri: gravatar.avatarForEmail(user.email) }}
              />
            </View>
            <Text
              onPress={this.openPrompt}
              style={styles.inputInstructions}
            >{user.displayName}
            </Text>
          </View>
        </View>
        <Text style={styles.usernameHeader}>User Name</Text>
        <Text style={styles.username}>{user.username}</Text>
        <Text style={styles.emailHeader}>Email Address</Text>
        <Text style={styles.email}>{user.email}</Text>
      </Container>
    );
  }
}

Profile.propTypes = {
  loading: PropTypes.bool,
  onChangeName: PropTypes.func,
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    email: PropTypes.string,
    displayName: PropTypes.string.isRequired,
  }),
};

export default Profile;
