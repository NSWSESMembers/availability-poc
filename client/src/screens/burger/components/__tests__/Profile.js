import React from 'react';
import TestRenderer from 'react-test-renderer';

import Profile from '../Profile';

const user = {
  username: 'username',
  email: 'email',
  displayName: 'displayName',
};

test('renders correctly', () => {
  const tree1 = TestRenderer.create(
    <Profile
      loading
      onChangeName={jest.fn()}
      user={user}
    />,
  ).toJSON();
  expect(tree1).toMatchSnapshot();

  const tree2 = TestRenderer.create(
    <Profile
      onChangeName={jest.fn()}
      user={user}
    />,
  ).toJSON();
  expect(tree2).toMatchSnapshot();
});

test('prompt', () => {
  const handleChangeName = jest.fn();
  const newName = 'test';

  const renderer = TestRenderer.create(
    <Profile
      onChangeName={handleChangeName}
      user={user}
    />,
  );

  const profile = renderer.getInstance();

  profile.openPrompt();
  profile.closePrompt();
  profile.updateDisplayName(newName);

  expect(handleChangeName.mock.calls.length).toBe(1);
  expect(handleChangeName.mock.calls[0][0]).toEqual(newName);
});
