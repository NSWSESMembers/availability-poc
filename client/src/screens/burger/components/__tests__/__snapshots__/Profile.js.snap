// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`renders correctly 2`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#EEE",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.8)",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "backgroundColor": "white",
              "borderRadius": 5,
              "borderWidth": 1,
              "elevation": 5,
              "marginTop": 150,
              "overflow": "hidden",
              "width": 300,
            },
            Object {
              "borderColor": "#ccc",
            },
            Object {},
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "borderBottomWidth": 1,
                "paddingHorizontal": 15,
                "paddingVertical": 10,
              },
              Object {
                "borderColor": "#ccc",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "fontSize": 18,
                  "fontWeight": "600",
                },
                Object {},
              ]
            }
          >
            Change Display Name
          </Text>
        </View>
        <View
          style={
            Object {
              "paddingHorizontal": 10,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            autoFocus={true}
            defaultValue="displayName"
            onChangeText={[Function]}
            placeholder="displayName"
            style={
              Array [
                Object {
                  "fontSize": 18,
                  "height": 50,
                },
                Object {},
              ]
            }
            underlineColorAndroid="white"
          />
        </View>
        <View
          style={
            Array [
              Object {
                "borderTopWidth": 1,
                "flexDirection": "row",
              },
              Object {
                "borderColor": "#ccc",
              },
            ]
          }
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "flex": 1,
                  "padding": 15,
                },
                Object {},
                Object {},
              ]
            }
            testID={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#006dbf",
                    "fontSize": 18,
                    "textAlign": "center",
                  },
                  Object {},
                  Object {},
                ]
              }
            >
              Cancel
            </Text>
          </View>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "flex": 1,
                  "padding": 15,
                },
                Object {},
                Object {},
              ]
            }
            testID={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#006dbf",
                    "fontSize": 18,
                    "textAlign": "center",
                  },
                  Object {},
                  Object {},
                ]
              }
            >
              OK
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>
  <View
    style={
      Object {
        "backgroundColor": "#c6c0c0",
        "paddingHorizontal": 16,
        "paddingLeft": 16,
        "paddingVertical": 8,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#dbdbdb",
          "flexDirection": "row",
          "paddingRight": 0,
          "paddingVertical": 8,
        }
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "https://www.gravatar.com/avatar/0c83f57c786a0b4a39efab23731c7ebc?d=mm",
            }
          }
          style={
            Object {
              "borderRadius": 5,
              "height": 50,
              "paddingHorizontal": 10,
              "width": 50,
            }
          }
        />
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        onPress={[Function]}
        style={
          Object {
            "color": "#777",
            "flex": 1,
            "fontSize": 26,
          }
        }
      >
        displayName
      </Text>
    </View>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "backgroundColor": "#dbdbdb",
        "color": "#000000",
        "fontSize": 16,
        "paddingBottom": 2,
        "paddingHorizontal": 8,
        "paddingTop": 5,
      }
    }
  >
    User Name
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#777",
        "borderTopWidth": 1,
        "fontSize": 14,
        "paddingBottom": 2,
        "paddingHorizontal": 8,
        "paddingTop": 5,
      }
    }
  >
    username
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "backgroundColor": "#dbdbdb",
        "color": "#000000",
        "fontSize": 16,
        "paddingBottom": 2,
        "paddingHorizontal": 8,
        "paddingTop": 5,
      }
    }
  >
    Email Address
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#777",
        "borderTopWidth": 1,
        "fontSize": 14,
        "paddingBottom": 2,
        "paddingHorizontal": 8,
        "paddingTop": 5,
      }
    }
  >
    email
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "backgroundColor": "#dbdbdb",
        "color": "#000000",
        "fontSize": 16,
        "paddingBottom": 2,
        "paddingHorizontal": 8,
        "paddingTop": 5,
      }
    }
  >
    Tags
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#777",
        "borderTopWidth": 1,
        "fontSize": 14,
        "paddingBottom": 2,
        "paddingHorizontal": 8,
        "paddingTop": 5,
      }
    }
  >
    tag1, tag2
  </Text>
</View>
`;
